src/cmdc/datasets/uscensus/data.py

```diff
@@ -0,0 +1,198 @@
+          ON vt.census_id=tt.census_id AND
+            vt.year={self.year} AND
+            vt.product='{self.product}'
+        ON CONFLICT {pk} DO UPDATE set value = excluded.value;
+        """
+
+        return _sql_data_insert
+
+    def data_get(self):
+        """
+        Fetches the data for the variables provided in the `__init__`
+        method from the specified ACS dataset
+
+        Returns
+        -------
+        df : pd.DataFrame
+            A DataFrame with a column for each variable requested and
+            a column, `fips` which specifies the geographic information
+        """
+        # Fetch data
+        df = super(ACS_CMDC, self).get(self.cols, self.geo)
+
+        # Convert to fips representation
+        df = self._create_fips(df)
+
+        # Reshape into desired format
+        df = df.melt(id_vars="fips", var_name="census_id", value_name="value")
+
+        return df
+
+    def _var_insert_query(self, df, table_name, temp_name, pk):
+        _sql_var_insert = f"""
+        INSERT INTO uscensus.{table_name} (year, product, census_id, label)
+        SELECT year, product, census_id, label FROM {temp_name}
+        ON CONFLICT {pk} DO NOTHING;
+        """
+
+        return _sql_var_insert
+
+    def variable_get(self):
+        """
+        Fetches the variables for the specified ACS dataset
+
+        Returns
+        -------
+        all_variables : pd.DataFrame
+            A DataFrame with columns (year, product, census_id, label)
+            which matches the columsn in the `uscensus.acs_variables`
+            table
+        """
+        # Fetch variables json
+        variable_json = json.loads(
+            requests.get(self.dataset["c_variablesLink"]).text
+        )
+
+        # Load into DataFrame
+        all_variables = pd.DataFrame.from_dict(
+            variable_json["variables"]
+        ).T
+
+        # Only keep 'Estimate' variables
+        is_variable = all_variables["label"].str.contains("Estimate")
+        all_variables = all_variables.loc[is_variable, ["label"]].reset_index()
+
+        all_variables["year"] = self.dataset["c_vintage"]
+        all_variables["product"] = self.product
+        all_variables = all_variables.rename(columns={"index": "census_id"})
+
+        return all_variables
+
+    def _put(self, connstr, table="variable", df=None):
+        """
+        Inserts the variable data into the `uscensus.acs_variables`
+        table
+
+        Parameters
+        ----------
+        connstr : string
+            A postgres connection string that can be used by sqlalchemy
+        d_or_v : string
+            A string that denotes whether we're putting data or
+            variables into the database
+        df : pd.DataFrame
+            The DataFrame with all of the variables contained in a
+            particular data product
+        """
+        if df is None:
+            if table == "data":
+                df = self.variable_get()
```

Comment:
maybe this should be `df = self.data_get()`?

Reply:
:astonished: